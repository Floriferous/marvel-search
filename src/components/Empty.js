import React from 'react';
import PropTypes from 'prop-types';

const Empty = ({ isSearching }) => (
  <h2 className="empty">
    {isSearching
      ? "Couln't find any results for this query"
      : 'Start searching or add bookmarks to find your heroes'}
  </h2>
);

Empty.propTypes = {
  isSearching: PropTypes.bool.isRequired,
};

export default Empty;
