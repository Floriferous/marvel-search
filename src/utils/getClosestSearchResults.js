import { createSearchKey } from '../actions/search';

/**
 * getClosestSearchResults - Searches for the closest search results
 * currently in memory. It iterates on a search string, and removes one
 * character at a time until it finds an old search result to display.
 *
 * If the first run passes, it simply returns the exact search results.
 *
 * An isLoading flag is also returned, to indicate whether old or exact
 * results are currently being returned
 *
 * @param {Object} searchResults Object with all previous search results
 * @param {String} search        current search string
 * @param {String} pagination    current pagination
 *
 * @return {Object}
 */
const getClosestSearchResults = (searchResults, search, pagination) => {
  let broaderResults;
  let isExactSearch;

  // Return immediately if search length equals 1 or pagination is larger than 0
  if (search.length === 1 || pagination > 0) {
    return {
      data: searchResults[createSearchKey(search, pagination)] || {},
      isLoading: false, // taken care of by CharacterList
    };
  }

  for (let i = 0; i < search.length; i += 1) {
    const broaderSearch = search.slice(0, search.length - i);
    broaderResults = searchResults[createSearchKey(broaderSearch, pagination)];
    if (broaderResults) {
      isExactSearch = i === 0;
      break;
    }
  }
  return { data: broaderResults || {}, isLoading: !isExactSearch };
};

export default getClosestSearchResults;
