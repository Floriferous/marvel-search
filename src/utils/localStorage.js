export const storageAvailable = () => {
  try {
    const storage = localStorage;
    const x = '__storage_test__';
    storage.setItem(x, x);
    storage.removeItem(x);
    return true;
  } catch (e) {
    return false;
  }
};

export const saveValueAtKey = (key, value) => {
  if (storageAvailable()) {
    localStorage.setItem(key, JSON.stringify(value));
    return true;
  }

  return false;
};

export const getValueAtKey = (key) => {
  if (storageAvailable()) {
    return JSON.parse(localStorage.getItem(key));
  }

  return false;
};
