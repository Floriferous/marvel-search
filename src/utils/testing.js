import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

export const testComponent = (Component, props) => {
  if (!testComponent.mountedLockScreen) {
    testComponent.mountedLockScreen = Enzyme.shallow(<Component {...props} />);
  }
  return testComponent.mountedLockScreen;
};

testComponent.reset = () => {
  testComponent.mountedLockScreen = undefined;
};

export const { shallow, mount, render } = Enzyme;
export default Enzyme;
