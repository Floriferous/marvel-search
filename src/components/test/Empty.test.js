import { testComponent } from '../../utils/testing';
import Empty from '../Empty';

describe('Empty', () => {
  let props;
  const component = () => testComponent(Empty, props);

  beforeEach(() => {
    props = { isSearching: false };
    testComponent.reset();
  });

  it('matches its snapshot', () => {
    expect(component()).toMatchSnapshot();
  });

  it('renders a h2 with class helper', () => {
    expect(component().find('h2').length).toBe(1);
    expect(component()
      .find('h2')
      .hasClass('empty')).toBe(true);
  });

  it('renders different test based on isBookmarked prop', () => {
    const text1 = component()
      .find('h2')
      .text();
    testComponent.reset();
    props.isSearching = true;
    const text2 = component()
      .find('h2')
      .text();

    expect(text1.length).toBeGreaterThan(0);
    expect(text2.length).toBeGreaterThan(0);
    expect(text1).not.toBe(text2);
  });
});
