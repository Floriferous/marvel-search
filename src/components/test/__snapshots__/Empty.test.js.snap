// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty matches its snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Empty
    isSearching={false}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": "Start searching or add bookmarks to find your heroes",
      "className": "empty",
    },
    "ref": null,
    "rendered": "Start searching or add bookmarks to find your heroes",
    "type": "h2",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": "Start searching or add bookmarks to find your heroes",
        "className": "empty",
      },
      "ref": null,
      "rendered": "Start searching or add bookmarks to find your heroes",
      "type": "h2",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
